README.md
# cryptohash-md5  [![Build Status](https://travis-ci.org/hvr/cryptohash-md5.svg?branch=master)](https://travis-ci.org/hvr/cryptohash-md5)

A practical incremental and one-pass, pure API to the [MD5 hash
algorithm](https://en.wikipedia.org/wiki/MD5) (including
[HMAC](https://en.wikipedia.org/wiki/HMAC) support) with performance
close to the fastest implementations available in other languages.


changelog.md
## 0.11.101.0

 - Add Eq instance for Ctx
 - add start and startlazy producing Ctx
 - Remove ineffective RULES

## 0.11.100.1

 - Use `__builtin_bswap{32,64}` only with GCC >= 4.3

## 0.11.100.0

 - new `hmac` and `hmaclazy` functions providing HMAC-MD5
   computation conforming to RFC2104 and RFC2202

## 0.11.7.2

 - switch to 'safe' FFI for calls where overhead becomes neglible
 - removed inline assembly in favour of portable C constructs
 - fix 32bit length overflow bug in `hash` function
 - fix inaccurate context-size
 - add context-size verification to incremental API operations
 - fix unaligned memory-accesses

## 0.11.7.1

 - first version forked off `cryptohash-0.11.7` release


Tests/KAT.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ViewPatterns #-}
import Data.Char
import Data.Bits
import Data.Word
import Data.ByteString (ByteString)
import Data.Foldable (foldl')
import Data.Monoid (mconcat)
import qualified Data.ByteString as B
import qualified Data.ByteString.Char8 as BC
import qualified Crypto.Hash.MD5 as MD5

import Test.Tasty
import Test.Tasty.QuickCheck
import Test.Tasty.HUnit

v0,v1,v2 :: ByteString
v0 = ""
v1 = "The quick brown fox jumps over the lazy dog"
v2 = "The quick brown fox jumps over the lazy cog"
vectors = [ v0, v1, v2 ]

instance Arbitrary ByteString where
    arbitrary = B.pack `fmap` arbitrary

data HashFct = HashFct
    { fctHash   :: (B.ByteString -> B.ByteString)
    , fctInc    :: ([B.ByteString] -> B.ByteString) }

hashinc i u f = f . foldl u i

md5Hash = HashFct { fctHash = MD5.hash, fctInc = hashinc MD5.init MD5.update MD5.finalize }


results :: [ (String, HashFct, [String]) ]
results = [
    ("MD5", md5Hash, [
        "d41d8cd98f00b204e9800998ecf8427e",
        "9e107d9d372bb6826bd81d3542a419d6",
        "1055d3e698d289f2af8663725127bd4b" ])
    ]

hexalise s = concatMap (\c -> [ hex $ c `div` 16, hex $ c `mod` 16 ]) s
        where hex i
                | i >= 0 && i <= 9   = fromIntegral (ord '0') + i
                | i >= 10 && i <= 15 = fromIntegral (ord 'a') + i - 10
                | otherwise          = 0

hexaliseB :: B.ByteString -> B.ByteString
hexaliseB = B.pack . hexalise . B.unpack

splitB :: Int -> ByteString -> [ByteString]
splitB l b =
    if B.length b > l
        then
            let (b1, b2) = B.splitAt l b in
            b1 : splitB l b2
        else
            [ b ]

showHash :: B.ByteString -> String
showHash = map (toEnum.fromEnum) . hexalise . B.unpack

runhash hash v = showHash $ (fctHash hash) $ v
runhashinc hash v = showHash $ (fctInc hash) $ v

makeTestAlg (name, hash, results) = testGroup name $ concatMap maketest (zip3 [0..] vectors results)
    where
        runtest :: ByteString -> String
        runtest v = runhash hash v

        runtestinc :: Int -> ByteString -> String
        runtestinc i v = runhashinc hash $ splitB i v

        maketest (i, v, r) =
            [ testCase (show i ++ " one-pass") (r @=? runtest v)
            , testCase (show i ++ " inc 1") (r @=? runtestinc 1 v)
            , testCase (show i ++ " inc 2") (r @=? runtestinc 2 v)
            , testCase (show i ++ " inc 3") (r @=? runtestinc 3 v)
            , testCase (show i ++ " inc 4") (r @=? runtestinc 4 v)
            , testCase (show i ++ " inc 5") (r @=? runtestinc 5 v)
            , testCase (show i ++ " inc 9") (r @=? runtestinc 9 v)
            , testCase (show i ++ " inc 16") (r @=? runtestinc 16 v)
            ]

katTests :: [TestTree]
katTests = map makeTestAlg results


main = defaultMain $ testGroup "cryptohash"
    [ testGroup "KATs" katTests
    ]


cbits/md5.c
/*
 * Copyright (C) 2006-2009 Vincent Hanquez <vincent@snarc.org>
 *               2016      Herbert Valerio Riedel <hvr@gnu.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "md5.h"

#include <assert.h>
#include <string.h>
#include <ghcautoconf.h>

#if defined(static_assert)
static_assert(sizeof(struct md5_ctx) == MD5_CTX_SIZE, "unexpected md5_ctx size");
#else
/* poor man's pre-C11 _Static_assert */
typedef char static_assertion__unexpected_md5_ctx_size[(sizeof(struct md5_ctx) == MD5_CTX_SIZE)?1:-1];
#endif

#define ptr_uint32_aligned(ptr) (!((uintptr_t)(ptr) & 0x3))

static inline uint32_t
rol32(const uint32_t word, const unsigned shift)
{
  /* GCC usually transforms this into a 'rol'-insn */
  return (word << shift) | (word >> (32 - shift));
}

static inline uint32_t
cpu_to_le32(const uint32_t hl)
{
#if !WORDS_BIGENDIAN
  return hl;
#elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3)
  return __builtin_bswap32(hl);
#else
  /* GCC usually transforms this into a bswap insn */
  return ((hl & 0xff000000) >> 24) |
         ((hl & 0x00ff0000) >> 8)  |
         ((hl & 0x0000ff00) << 8)  |
         ( hl               << 24);
#endif
}

static inline void
cpu_to_le32_array(uint32_t *dest, const uint32_t *src, unsigned wordcnt)
{
  while (wordcnt--)
    *dest++ = cpu_to_le32(*src++);
}

static inline uint64_t
cpu_to_le64(const uint64_t hll)
{
#if !WORDS_BIGENDIAN
  return hll;
#elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 3)
  return __builtin_bswap64(hll);
#else
  return ((uint64_t)cpu_to_le32(hll & 0xffffffff) << 32LL) | cpu_to_le32(hll >> 32);
#endif
}


void
hs_cryptohash_md5_init(struct md5_ctx *ctx)
{
  memset(ctx, 0, sizeof(*ctx));

  ctx->h[0] = 0x67452301;
  ctx->h[1] = 0xefcdab89;
  ctx->h[2] = 0x98badcfe;
  ctx->h[3] = 0x10325476;
}

#define f1(x, y, z)     (z ^ (x & (y ^ z)))
#define f2(x, y, z)     f1(z, x, y)
#define f3(x, y, z)     (x ^ y ^ z)
#define f4(x, y, z)     (y ^ (x | ~z))
#define R(f, a, b, c, d, i, k, s) a += f(b, c, d) + w[i] + k; a = rol32(a, s); a += b

static void
md5_do_chunk_aligned(struct md5_ctx *ctx, const uint32_t w[])
{
  uint32_t a = ctx->h[0];
  uint32_t b = ctx->h[1];
  uint32_t c = ctx->h[2];
  uint32_t d = ctx->h[3];

  R(f1, a, b, c, d,  0, 0xd76aa478,  7);
  R(f1, d, a, b, c,  1, 0xe8c7b756, 12);
  R(f1, c, d, a, b,  2, 0x242070db, 17);
  R(f1, b, c, d, a,  3, 0xc1bdceee, 22);
  R(f1, a, b, c, d,  4, 0xf57c0faf,  7);
  R(f1, d, a, b, c,  5, 0x4787c62a, 12);
  R(f1, c, d, a, b,  6, 0xa8304613, 17);
  R(f1, b, c, d, a,  7, 0xfd469501, 22);
  R(f1, a, b, c, d,  8, 0x698098d8,  7);
  R(f1, d, a, b, c,  9, 0x8b44f7af, 12);
  R(f1, c, d, a, b, 10, 0xffff5bb1, 17);
  R(f1, b, c, d, a, 11, 0x895cd7be, 22);
  R(f1, a, b, c, d, 12, 0x6b901122,  7);
  R(f1, d, a, b, c, 13, 0xfd987193, 12);
  R(f1, c, d, a, b, 14, 0xa679438e, 17);
  R(f1, b, c, d, a, 15, 0x49b40821, 22);

  R(f2, a, b, c, d,  1, 0xf61e2562,  5);
  R(f2, d, a, b, c,  6, 0xc040b340,  9);
  R(f2, c, d, a, b, 11, 0x265e5a51, 14);
  R(f2, b, c, d, a,  0, 0xe9b6c7aa, 20);
  R(f2, a, b, c, d,  5, 0xd62f105d,  5);
  R(f2, d, a, b, c, 10, 0x02441453,  9);
  R(f2, c, d, a, b, 15, 0xd8a1e681, 14);
  R(f2, b, c, d, a,  4, 0xe7d3fbc8, 20);
  R(f2, a, b, c, d,  9, 0x21e1cde6,  5);
  R(f2, d, a, b, c, 14, 0xc33707d6,  9);
  R(f2, c, d, a, b,  3, 0xf4d50d87, 14);
  R(f2, b, c, d, a,  8, 0x455a14ed, 20);
  R(f2, a, b, c, d, 13, 0xa9e3e905,  5);
  R(f2, d, a, b, c,  2, 0xfcefa3f8,  9);
  R(f2, c, d, a, b,  7, 0x676f02d9, 14);
  R(f2, b, c, d, a, 12, 0x8d2a4c8a, 20);

  R(f3, a, b, c, d,  5, 0xfffa3942,  4);
  R(f3, d, a, b, c,  8, 0x8771f681, 11);
  R(f3, c, d, a, b, 11, 0x6d9d6122, 16);
  R(f3, b, c, d, a, 14, 0xfde5380c, 23);
  R(f3, a, b, c, d,  1, 0xa4beea44,  4);
  R(f3, d, a, b, c,  4, 0x4bdecfa9, 11);
  R(f3, c, d, a, b,  7, 0xf6bb4b60, 16);
  R(f3, b, c, d, a, 10, 0xbebfbc70, 23);
  R(f3, a, b, c, d, 13, 0x289b7ec6,  4);
  R(f3, d, a, b, c,  0, 0xeaa127fa, 11);
  R(f3, c, d, a, b,  3, 0xd4ef3085, 16);
  R(f3, b, c, d, a,  6, 0x04881d05, 23);
  R(f3, a, b, c, d,  9, 0xd9d4d039,  4);
  R(f3, d, a, b, c, 12, 0xe6db99e5, 11);
  R(f3, c, d, a, b, 15, 0x1fa27cf8, 16);
  R(f3, b, c, d, a,  2, 0xc4ac5665, 23);

  R(f4, a, b, c, d,  0, 0xf4292244,  6);
  R(f4, d, a, b, c,  7, 0x432aff97, 10);
  R(f4, c, d, a, b, 14, 0xab9423a7, 15);
  R(f4, b, c, d, a,  5, 0xfc93a039, 21);
  R(f4, a, b, c, d, 12, 0x655b59c3,  6);
  R(f4, d, a, b, c,  3, 0x8f0ccc92, 10);
  R(f4, c, d, a, b, 10, 0xffeff47d, 15);
  R(f4, b, c, d, a,  1, 0x85845dd1, 21);
  R(f4, a, b, c, d,  8, 0x6fa87e4f,  6);
  R(f4, d, a, b, c, 15, 0xfe2ce6e0, 10);
  R(f4, c, d, a, b,  6, 0xa3014314, 15);
  R(f4, b, c, d, a, 13, 0x4e0811a1, 21);
  R(f4, a, b, c, d,  4, 0xf7537e82,  6);
  R(f4, d, a, b, c, 11, 0xbd3af235, 10);
  R(f4, c, d, a, b,  2, 0x2ad7d2bb, 15);
  R(f4, b, c, d, a,  9, 0xeb86d391, 21);

  ctx->h[0] += a;
  ctx->h[1] += b;
  ctx->h[2] += c;
  ctx->h[3] += d;
}

static void
md5_do_chunk(struct md5_ctx *ctx, const uint8_t buf[])
{
  if (ptr_uint32_aligned(buf)) { /* aligned buf */
#if WORDS_BIGENDIAN
    uint32_t w[16]; cpu_to_le32_array(w, (const uint32_t *)buf, 16);
#else
    const uint32_t *w = (const uint32_t *)buf;
#endif
    md5_do_chunk_aligned(ctx, w);
  } else { /* unaligned buf */
    uint32_t w[16]; memcpy(w, buf, 64);
#if WORDS_BIGENDIAN
    cpu_to_le32_array(w, w, 16);
#endif
    md5_do_chunk_aligned(ctx, w);
  }
}

void
hs_cryptohash_md5_update(struct md5_ctx *ctx, const uint8_t *data, size_t len)
{
  size_t index = ctx->sz & 0x3f;
  const size_t to_fill = 64 - index;

  ctx->sz += len;

  /* process partial buffer if there's enough data to make a block */
  if (index && len >= to_fill) {
    memcpy(ctx->buf + index, data, to_fill);
    md5_do_chunk(ctx, ctx->buf);
    /* memset(ctx->buf, 0, 64); */
    len -= to_fill;
    data += to_fill;
    index = 0;
  }

  /* process as many 64-blocks as possible */
  while (len >= 64) {
    md5_do_chunk(ctx, data);
    len -= 64;
    data += 64;
  }

  /* append data into buf */
  if (len)
    memcpy(ctx->buf + index, data, len);
}

void
hs_cryptohash_md5_finalize(struct md5_ctx *ctx, uint8_t *out)
{
  static const uint8_t padding[64] = { 0x80, };

  /* add padding and update data with it */
  const uint64_t bits = cpu_to_le64(ctx->sz << 3);

  /* pad out to 56 */
  const size_t index = (ctx->sz & 0x3f);
  const size_t padlen = (index < 56) ? (56 - index) : ((64 + 56) - index);
  hs_cryptohash_md5_update(ctx, padding, padlen);

  /* append length */
  hs_cryptohash_md5_update(ctx, (const uint8_t *) &bits, sizeof(bits));

  /* output hash */
  cpu_to_le32_array((uint32_t *) out, ctx->h, 4);
}


cbits/md5.h
/*
 * Copyright (C) 2006-2009 Vincent Hanquez <vincent@snarc.org>
 *               2016      Herbert Valerio Riedel <hvr@gnu.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef CRYPTOHASH_MD5_H
#define CRYPTOHASH_MD5_H

#include <stdint.h>
#include <stddef.h>

struct md5_ctx
{
	uint64_t sz;
	uint8_t  buf[64];
	uint32_t h[4];
};

#define MD5_DIGEST_SIZE		16
#define MD5_CTX_SIZE		88

void hs_cryptohash_md5_init(struct md5_ctx *ctx);
void hs_cryptohash_md5_update(struct md5_ctx *ctx, const uint8_t *data, size_t len);
void hs_cryptohash_md5_finalize(struct md5_ctx *ctx, uint8_t *out);

#endif


Bench/Bench.hs
{-# LANGUAGE BangPatterns #-}
import Criterion.Main
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import qualified Crypto.Hash.MD5 as MD5

hashmany (i,u,f) = f . foldl u i

allHashs =
    [ ("MD5",MD5.hash, hashmany (MD5.init,MD5.update,MD5.finalize))
    ]

benchHash :: a -> (a -> B.ByteString) -> Benchmarkable
benchHash bs f = whnf f bs

withHashesFilter out f = map f $ filter (\(n,_,_) -> not (n `elem` out)) allHashs
withHashes f = map f allHashs

main = do
    let !bs32     = B.replicate 32 0
        !bs256    = B.replicate 256 0
        !bs4096   = B.replicate 4096 0
        !bs1M     = B.replicate (1*1024*1024) 0
    let !lbs64x256 = (map (const (B.replicate 64 0)) [0..3])
        !lbs64x4096 = (map (const (B.replicate 64 0)) [0..63])
    defaultMain
        [ bgroup "hash-32b" (withHashes (\(name, f,_) -> bench name $ benchHash bs32 f))
        , bgroup "hash-256b" (withHashes (\(name, f,_) -> bench name $ benchHash bs256 f))
        , bgroup "hash-4Kb" (withHashes (\(name, f,_) -> bench name $ benchHash bs4096 f))
        , bgroup "hash-1Mb" (withHashesFilter ["MD2"] (\(name, f,_) -> bench name $ benchHash bs1M f))
        , bgroup "iuf-64x256" (withHashes (\(name, _,f) -> bench name $ benchHash lbs64x256 f))
        , bgroup "iuf-64x4096" (withHashes (\(name, _,f) -> bench name $ benchHash lbs64x4096 f))
        ]


Bench/BenchAPI.hs
{-# LANGUAGE BangPatterns #-}
import Criterion.Main
import qualified Data.ByteString as B
import qualified Data.ByteString.Lazy as L
import qualified Crypto.Hash.MD5 as MD5

md5F = ( "MD5"
        , MD5.hash
        , MD5.finalize . MD5.update MD5.init
        )

main = do
    let !bs32     = B.replicate 32 0
        !bs256    = B.replicate 256 0
        !bs4096   = B.replicate 4096 0
        !bs1M     = B.replicate (1*1024*1024) 0
    let !lbs64x256 = (map (const (B.replicate 64 0)) [0..3])
        !lbs64x4096 = (map (const (B.replicate 64 0)) [0..63])

    let (fname, fHash, fIncr) = md5F
    let benchName ty z = fname ++ "." ++ ty -- ++ " " ++ show z
    defaultMain
        [ bgroup "hash-0b"
            [ bench (benchName "hash" 0) $ whnf fHash B.empty
            , bench (benchName "incr" 0) $ whnf fIncr B.empty
            ]
        , bgroup "hash-32b"
            [ bench (benchName "hash" 32) $ whnf fHash bs32
            , bench (benchName "incr" 32) $ whnf fIncr bs32
            ]
        , bgroup "hash-256b"
            [ bench (benchName "hash" 256) $ whnf fHash bs256
            , bench (benchName "incr" 256) $ whnf fIncr bs256
            ]
        , bgroup "hash-4Kb"
            [ bench (benchName "hash" 4096) $ whnf fHash bs4096
            , bench (benchName "incr" 4096) $ whnf fIncr bs4096
            ]
        ]


src-tests/test-md5.hs
{-# LANGUAGE OverloadedStrings #-}

module Main (main) where

import           Data.Bits              (xor)
import           Data.ByteString        (ByteString)
import qualified Data.ByteString        as B
import qualified Data.ByteString.Lazy   as BL
import qualified Data.ByteString.Base16 as B16

-- reference implementation
import qualified Data.Digest.Pure.MD5 as REF

-- implementation under test
import qualified Crypto.Hash.MD5      as IUT

import           Test.Tasty
import           Test.Tasty.HUnit
import           Test.Tasty.QuickCheck as QC

vectors :: [ByteString]
vectors =
    [ ""
    , "The quick brown fox jumps over the lazy dog"
    , "The quick brown fox jumps over the lazy cog"
    , "abc"
    , "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
    , "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmnoijklmnopjklmnopqklmnopqrlmnopqrsmnopqrstnopqrstu"
    , B.replicate 1000000 0x61
    ]

answers :: [ByteString]
answers = map (B.filter (/= 0x20))
    [ "d41d8cd9 8f00b204 e9800998 ecf8427e"
    , "9e107d9d 372bb682 6bd81d35 42a419d6"
    , "1055d3e6 98d289f2 af866372 5127bd4b"
    , "90015098 3cd24fb0 d6963f7d 28e17f72"
    , "8215ef07 96a20bca aae116d3 876c664a"
    , "03dd8807 a93175fb 062dfb55 dc7d359c"
    , "7707d6ae 4e027c70 eea2a935 c2296f21"
    ]

ansXLTest :: ByteString
ansXLTest = B.filter (/= 0x20)
    "d3381391 69d50f55 526194c7 90ec0448"

katTests :: [TestTree]
katTests
  | length vectors == length answers = map makeTest (zip3 [1::Int ..] vectors answers) ++ [xltest, xltest']
  | otherwise = error "vectors/answers length mismatch"
  where
    makeTest (i, v, r) = testGroup ("vec"++show i) $
        [ testCase "one-pass" (r @=? runTest v)
        , testCase "one-pass'" (r @=? runTest' v)
        , testCase "inc-1"    (r @=? runTestInc 1 v)
        , testCase "inc-2"    (r @=? runTestInc 2 v)
        , testCase "inc-3"    (r @=? runTestInc 3 v)
        , testCase "inc-4"    (r @=? runTestInc 4 v)
        , testCase "inc-5"    (r @=? runTestInc 5 v)
        , testCase "inc-7"    (r @=? runTestInc 7 v)
        , testCase "inc-8"    (r @=? runTestInc 8 v)
        , testCase "inc-9"    (r @=? runTestInc 9 v)
        , testCase "inc-16"   (r @=? runTestInc 16 v)
        , testCase "lazy-1"   (r @=? runTestLazy 1 v)
        , testCase "lazy-2"   (r @=? runTestLazy 2 v)
        , testCase "lazy-7"   (r @=? runTestLazy 7 v)
        , testCase "lazy-8"   (r @=? runTestLazy 8 v)
        , testCase "lazy-16"  (r @=? runTestLazy 16 v)
        , testCase "lazy-1'"   (r @=? runTestLazy' 1 v)
        , testCase "lazy-2'"   (r @=? runTestLazy' 2 v)
        , testCase "lazy-7'"   (r @=? runTestLazy' 7 v)
        , testCase "lazy-8'"   (r @=? runTestLazy' 8 v)
        , testCase "lazy-16'"  (r @=? runTestLazy' 16 v)
        ] ++
        [ testCase "lazy-63u"  (r @=? runTestLazyU 63 v) | B.length v > 63 ] ++
        [ testCase "lazy-65u"  (r @=? runTestLazyU 65 v) | B.length v > 65 ] ++
        [ testCase "lazy-97u"  (r @=? runTestLazyU 97 v) | B.length v > 97 ] ++
        [ testCase "lazy-131u" (r @=? runTestLazyU 131 v) | B.length v > 131 ] ++
        [ testCase "lazy-63u'"  (r @=? runTestLazyU' 63 v) | B.length v > 63 ] ++
        [ testCase "lazy-65u'"  (r @=? runTestLazyU' 65 v) | B.length v > 65 ] ++
        [ testCase "lazy-97u'"  (r @=? runTestLazyU' 97 v) | B.length v > 97 ] ++
        [ testCase "lazy-131u'" (r @=? runTestLazyU' 131 v) | B.length v > 131 ]

    runTest :: ByteString -> ByteString
    runTest = B16.encode . IUT.hash

    runTest' :: ByteString -> ByteString
    runTest' = B16.encode . IUT.finalize . IUT.start

    runTestInc :: Int -> ByteString -> ByteString
    runTestInc i = B16.encode . IUT.finalize . myfoldl' IUT.update IUT.init . splitB i

    runTestLazy :: Int -> ByteString -> ByteString
    runTestLazy i = B16.encode . IUT.hashlazy . BL.fromChunks . splitB i

    runTestLazy' :: Int -> ByteString -> ByteString
    runTestLazy' i = B16.encode . IUT.finalize . IUT.startlazy . BL.fromChunks . splitB i

    -- force unaligned md5-blocks
    runTestLazyU :: Int -> ByteString -> ByteString
    runTestLazyU i = B16.encode . IUT.hashlazy . BL.fromChunks . map B.copy . splitB i

    runTestLazyU' :: Int -> ByteString -> ByteString
    runTestLazyU' i = B16.encode . IUT.finalize . IUT.startlazy . BL.fromChunks . map B.copy . splitB i

    ----

    xltest = testGroup "XL-vec"
        [ testCase "inc" (ansXLTest @=? (B16.encode . IUT.hashlazy) vecXL) ]
      where
        vecXL = BL.fromChunks (replicate 16777216 "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno")

    xltest' = testGroup "XL-vec'"
        [ testCase "inc'" (ansXLTest @=? (B16.encode . IUT.finalize . IUT.startlazy) vecXL) ]
      where
        vecXL = BL.fromChunks (replicate 16777216 "abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno")

splitB :: Int -> ByteString -> [ByteString]
splitB l b
  | B.length b > l = b1 : splitB l b2
  | otherwise = [b]
  where
    (b1, b2) = B.splitAt l b


rfc2202Vectors :: [(ByteString,ByteString,ByteString)]
rfc2202Vectors = -- (secrect,msg,mac)
    [ (rep 16 0x0b, "Hi There", x"9294727a3638bb1c13f48ef8158bfc9d")
    , ("Jefe", "what do ya want for nothing?", x"750c783e6ab0b503eaa86e310a5db738")
    , (rep 16 0xaa, rep 50 0xdd, x"56be34521d144c88dbb8c733f0e8b3f6")
    , (B.pack [1..25], rep 50 0xcd, x"697eaf0aca3a3aea3a75164746ffaa79")
    , (rep 16 0x0c, "Test With Truncation", x"56461ef2342edc00f9bab995690efd4c")
    , (rep 80 0xaa, "Test Using Larger Than Block-Size Key - Hash Key First", x"6b1ab7fe4bd7bf8f0b62e6ce61b9d0cd")
    , (rep 80 0xaa, "Test Using Larger Than Block-Size Key and Larger Than One Block-Size Data", x"6f630fad67cda0ee1fb1f562db3aa53e")
    ]
  where
    x = fst.B16.decode
    rep n c = B.replicate n c

rfc2202Tests :: [TestTree]
rfc2202Tests = zipWith makeTest [1::Int ..] rfc2202Vectors
  where
    makeTest i (key, msg, mac) = testGroup ("vec"++show i) $
        [ testCase "hmac" (hex mac  @=? hex (IUT.hmac key msg))
        , testCase "hmaclazy" (hex mac  @=? hex (IUT.hmaclazy key lazymsg))
        ]
      where
        lazymsg = BL.fromChunks . splitB 1 $ msg

    hex = B16.encode

-- define own 'foldl' here to avoid RULE rewriting to 'startlazy
myfoldl' :: (b -> a -> b) -> b -> [a] -> b
myfoldl' f z0 xs0 = lgo z0 xs0
  where
    lgo z []     = z
    lgo z (x:xs) = let z' = f z x
                   in z' `seq` lgo z' xs

newtype RandBS = RandBS { unRandBS :: ByteString }
newtype RandLBS = RandLBS BL.ByteString

instance Arbitrary RandBS where
    arbitrary = fmap (RandBS . B.pack) arbitrary
    shrink (RandBS x) = fmap RandBS (go x)
      where
        go bs = zipWith B.append (B.inits bs) (tail $ B.tails bs)

instance Show RandBS where
    show (RandBS x) = "RandBS {len=" ++ show (B.length x)++"}"

instance Arbitrary RandLBS where
    arbitrary = fmap (RandLBS . BL.fromChunks . map unRandBS) arbitrary

instance Show RandLBS where
    show (RandLBS x) = "RandLBS {len=" ++ show (BL.length x) ++ ", chunks=" ++ show (length $ BL.toChunks x)++"}"

refImplTests :: [TestTree]
refImplTests =
    [ testProperty "hash" prop_hash
    , testProperty "start" prop_start
    , testProperty "hashlazy" prop_hashlazy
    , testProperty "startlazy" prop_startlazy
    , testProperty "hmac" prop_hmac
    , testProperty "hmaclazy" prop_hmaclazy
    ]
  where
    prop_hash (RandBS bs)
        = ref_hash bs == IUT.hash bs

    prop_start (RandBS bs)
        = ref_hash bs == (IUT.finalize $ IUT.start bs)

    prop_hashlazy (RandLBS bs)
        = ref_hashlazy bs == IUT.hashlazy bs

    prop_startlazy (RandLBS bs)
        = ref_hashlazy bs == (IUT.finalize $ IUT.startlazy bs)

    prop_hmac (RandBS k) (RandBS bs)
        = ref_hmac k bs == IUT.hmac k bs

    prop_hmaclazy (RandBS k) (RandLBS bs)
        = ref_hmaclazy k bs == IUT.hmaclazy k bs

    ref_hash :: ByteString -> ByteString
    ref_hash = ref_hashlazy . fromStrict

    ref_hashlazy :: BL.ByteString -> ByteString
    ref_hashlazy = REF.md5DigestBytes . REF.md5

    -- stolen & adapted from SHA package
    ref_hmac :: ByteString -> ByteString -> ByteString
    ref_hmac k m = ref_hash (B.append opad (ref_hash (B.append ipad m)))
     where
      opad = B.map (xor 0x5c) k'
      ipad = B.map (xor 0x36) k'

      k' = B.append kt pad
       where
        kt  = if kn > bn then ref_hash k else k
        pad = B.replicate (bn - ktn) 0
        kn  = B.length k
        ktn = B.length kt
        bn  = 64

    ref_hmaclazy :: ByteString -> BL.ByteString -> ByteString
    ref_hmaclazy secret = ref_hmac secret . toStrict

    -- toStrict/fromStrict only available with bytestring-0.10 and later
    toStrict = B.concat . BL.toChunks
    fromStrict = BL.fromChunks . (:[])

main :: IO ()
main = defaultMain $ testGroup "cryptohash-md5"
    [ testGroup "KATs" katTests
    , testGroup "RFC2202" rfc2202Tests
    , testGroup "REF" refImplTests
    ]
